web/packages/teleterm/src/ui/services/modals/modalsService.ts

```diff
@@ -58,20 +67,44 @@ export class ModalsService extends ImmutableStore<State> {
    * old dialog with the new one.
    * The old dialog is canceled, if possible.
    *
-   * The returned closeDialog function can be used to close the dialog and automatically call the
-   * dialog's onCancel callback (if present).
+   * The passed `abortSignal` or the returned `closeDialog` function can be used to close the dialog
+   * and automatically call the dialog's onCancel callback (if present).
    */
-  openRegularDialog(dialog: Dialog): { closeDialog: () => void } {
+  openRegularDialog(
+    dialog: Dialog,
+    abortSignal?: AbortSignal
+  ): {
+    closeDialog: () => void;
+  } {
+    const onCancelDialog = () => dialog['onCancel']?.();
+    const sharedSignal = this.getSharedAbortSignal(abortSignal);
+    if (sharedSignal.aborted) {
+      onCancelDialog();
+      return {
+        closeDialog: () => {},
+      };
+    }
+
     this.state.regular?.['onCancel']?.();
+
+    const closeDialog = () => {
+      sharedSignal.removeEventListener('abort', closeDialog);
+      // Do we still have the same dialog open?
+      if (dialog !== this.state.regular) {
+        return;
+      }
+
+      onCancelDialog();
+      this.closeRegularDialog();
```

Comment:
Any reason this order is reversed compared to the original where we first closed the dialog and only then executed its `onCancel` callback?

Reply:
No, reverted.